src/main/java/net/snowflake/client/jdbc/SnowflakeSQLLoggedException.java

```diff
@@ -0,0 +1,238 @@
+import net.snowflake.client.jdbc.telemetry.TelemetryField;
+import net.snowflake.client.jdbc.telemetryOOB.TelemetryEvent;
+import net.snowflake.client.jdbc.telemetryOOB.TelemetryService;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * @author mknister
+ *
+ * This SnowflakeSQLLoggedException class extends the SnowflakeSQLException class to add OOB telemetry data for sql
+ * exceptions. Not all sql exceptions require OOB telemetry logging so the exceptions in this class should only be
+ * thrown if there is a need for logging the exception with OOB telemetry.
+ */
+
+public class SnowflakeSQLLoggedException extends SnowflakeSQLException
+{
+  /*
+  OOB telemetry instance
+   */
+  private TelemetryService oobInstance = TelemetryService.getInstance();
+
+  /*
+  IB telemtry instance
+   */
+  private Telemetry ibInstance = null;
+
+  private final static ObjectMapper mapper =
+          ObjectMapperFactory.getObjectMapper();
+
+  /**
+   * Function to create a TelemetryEvent log from the JSONObject and exception and send it via OOB telemetry
+   *
+   * @param value JSONnode containing relevant information specific to the exception constructor that
+   *              should be included in the telemetry data, such as sqlState or vendorCode
+   * @param ex    The exception being thrown
+   */
+  private void sendOutOfBandTelemetryMessage(JSONObject value, SnowflakeSQLLoggedException ex)
+  {
+    TelemetryEvent.LogBuilder logBuilder = new TelemetryEvent.LogBuilder();
+    StringWriter sw = new StringWriter();
+    PrintWriter pw = new PrintWriter(sw);
+    ex.printStackTrace(pw);
+    String stackTrace = sw.toString();
+    value.put("Stacktrace", stackTrace);
+    TelemetryEvent log = logBuilder
+        .withName("Exception: " + ex.getMessage())
+        .withValue(value)
+        .build();
+    oobInstance.report(log);
+  }
+
+  /**
+   * Function to create a TelemetryClient log and send it via in-band telemetry
+   *
+   * @param value ObjectNode containing information specific to the exception constructor that should be included in
+   *              the telemetry log, such as SQLState or reason for the error
+   * @param ex The exception being thrown
+   * @return true if in-band telemetry log sent successfully or false if it did not
+   */
+  private boolean sendInBandTelemetryMessage(ObjectNode value, SnowflakeSQLLoggedException ex) {
+    StringWriter sw = new StringWriter();
+    PrintWriter pw = new PrintWriter(sw);
+    ex.printStackTrace(pw);
+    String stackTrace = sw.toString();
+    value.put("Stacktrace", stackTrace);
+    ibInstance.addLogToBatch(new TelemetryData(value, System.currentTimeMillis()));
+    Future<Boolean> success = ibInstance.sendBatchAsync();
+    try
+    {
+      // wait for 10 seconds at most before giving up
+      return success.get(10, TimeUnit.SECONDS);
```

Comment:
Can you make this async? The main thread should not wait for the IB to send. Also 10 seconds is too long and use a while loop to check.

Reply:
If we don't wait for a response from IB, how do we know whether or not to send OOB?

Reply:
The success.get() function acts as a while loop. It will return in less than 10 seconds if the value returns sooner. 10 seconds is the max. What do you propose as the max value?

Reply:
Aha, agree with you that the 10 seconds is the timeout for `success.get`. The main issue is that `success.get` is going to block the main thread which is not necessary. To avoid this, you can do something here https://github.com/snowflakedb/snowflake-jdbc/blob/c0ac922376088c2fb65f2e6a6c4b59a8c515d86d/src/main/java/net/snowflake/client/jdbc/telemetry/TelemetryClient.java#L252
So it can be a general feature to report IB errors to OOB. WDYT?

Reply:
Do you think there is any way to not block the main thread but keep the code in SnowflakeSQLLoggedException? I think it's too big of a change to link together OOB and IB for every possible instance. Also, IB and OOB use different types of logs and I'd have to parse through the IB log to create a new OOB log and send it. It might be missing some information.

Reply:
You can create an executor in your log class and use it to submit IB and handle OOB if it fails.
```{java}
ExecutorService executor = Executors.newSingleThreadExecutor();
executor.submit(() -> {
    String threadName = Thread.currentThread().getName();
    System.out.println("Hello " + threadName);
});
```

Reply:
Thanks! Added an executor thread.